make/cpp2asm_support.cmake

```diff
@@ -138,10 +138,15 @@ if (UNIX)
   # and doesn't produce any output, so we must use cpp for our .asm files.
   # we assume it's in the same dir.
   get_filename_component(compiler_path ${CMAKE_C_COMPILER} PATH)
-  find_program(CMAKE_CPP cpp HINTS "${compiler_path}" DOC "path to C preprocessor")
-  if (NOT CMAKE_CPP)
-    message(FATAL_ERROR "cpp is required to build")
-  endif (NOT CMAKE_CPP)
+  # Allow user to set C pre-processor from environment variable
+  if (DEFINED ENV{CPP})
```

Comment:
Can't we use CMAKE_CPP directly as the override switch, do we need a separate CPP (I'm not sure, just checking because if possible that might be easier and less code)?

Reply:
I chose CPP just for consistency with the existing methods of overriding similar variables, like CC and CXX. I can change to CMAKE_CPP which would make a typical script look something like:
```
export CC='gcc-7'
export CXX='g++-7'
export CMAKE_CPP='cpp-7'
```
rather than:
```
export CC='gcc-7'
export CXX='g++-7'
export CPP='cpp-7'
```
Which do we prefer?

Reply:
Keep CPP then!

Reply:
Ok

Reply:
Do we need an env var at all?  find_program makes a CMake cache variable, which means you can set it at config time.  E.g., "cmake -DCMAKE_CPP=/my/path/to/cpp <srcpath>".  Is that sufficient?

Reply:
I guess the problem is that for setting a CMake cache variable from the command line through runsuite.cmake today we don't have the `cacheappend` convenience parameter we have for package.cmake: that may be nice to add anyway.  Without it you'd have to put the define in a file and use preload+preload64 I suppose.  So the choice seems to be, add a `cacheappend` (to runsuite_common_pre.cmake which adds it to base_cache) and then just set the CMake variable directly (this would also let you set other variables this way, so it seems nice to have), or have this specific env var.

Reply:
Assad, once/if this question i resolved, do you have merging privileges or do you need me to merge?

Reply:
@hgreving2304 I have merging privileges, thanks.

Reply:
@derekbruening From my understanding of the approaches you propose (`cmake -D` and `cacheappend`), their limitation in this case seems to be that they are to some extent wired into the `cmake` command for different types of build and are not as flexible as envvars.

For example, we support 3 distros all of which have slightly different toolchain requirements. Part of the nightly run script:
```
if [ "${OS}" == "susetumbleweed" ]; then
    CC='gcc-7'
    CXX='g++-7'
    CPP='cpp-7'
fi

if [ "${OS}" == "ubuntu1604" ]; then
    CXX='g++-5'
fi

docker run \
     -v `pwd`/dynamorio-src:/home/dynamorio/dynamorio-src \
     -v `pwd`/run_nightly_tests.sh:/tmp/run_nightly_tests.sh \
     -e CXX=${CXX} -e CC=${CC} -e CPP=${CPP} -e OS_STRING=${OS} \
     dynamorio-test/${OS} \
     /tmp/run_nightly_tests.sh &> ${OS}_run.log
```
The envvar approach above is simpler because the required non-default settings are propogated through to the `cmake` calls without any extra changes to the `cmake` command or the layers in between.

Perhaps it would be better to classify cmake config settings and the way they are set based on their use/intent?
For example, platform dependent toolchain settings should be treated and set differently from DynamoRIO build types, because build types are less dependent on platform types than toolchains.

Reply:
> The envvar approach above is simpler because the required non-default settings are propogated through to the build without any extra changes to the cmake command or the layers in between.

Right, you would have to pass parameters to `run_nightly_tests.sh` which would then pass them on to the cmake scripts in order to use the cmake config approach; or, you would make a toolchain cmake file containing your paths, much like you'd do for cross-compiling, and point at that file.

I am not opposed to adding this CPP env var.  However I'm not sure it's the right general direction if there will be more of these -- if we end up wanting to add a LINKER and ASSEMBLER and further settings I would vote against a separate env var for each of those: better to have a CMake toolchain file that sets CMake variables, just like with cross-compiling, if we have a lot of paths to specify.

> Perhaps it would be better to classify cmake config settings and the way they are set based on their use/intent?
> For example, platform dependent toolchain settings should be treated and set differently from DynamoRIO build types, because build types are less dependent on platform types than toolchains.

I'm not sure I'm following the proposal: could you elaborate on what coding changes you are suggesting?  CMake usage in general has always combined those types of settings: both toolchain paths and build types are specified via the same types of variables.

Reply:
> I'm not sure I'm following the proposal: could you elaborate on what coding changes

Sorry, I didn't mean to suggest coding changes. It was just policy or guidelines for deciding what is allowed to be passed as an envvar. 

> I am not opposed to adding this CPP env var. However I'm not sure it's the right general direction if there will be more of these

I agree it's not the right general direction. This is one case ATM. If it becomes a recurring pattern and the list of envvar overrides increases, then your suggestion of setting variables in CMake toolchain files will be a better solution. Ideally we should remove the need for these altogether by fixing the errors caused by later versions of the compiler.